tests/Unit/DataStructures/Test_DynamicBuffer.cpp

```diff
@@ -0,0 +1,124 @@
+    CHECK_ITERABLE_APPROX(*it, expected[i]);
+  }
+
+  i = 0;
+  for (const auto& val : dynamic_buffer) {
+    CHECK_ITERABLE_APPROX(val, expected[i]);
+    ++i;
+  }
+}
+
+template <typename T>
+std::vector<T> create_random_data(const gsl::not_null<std::mt19937*> gen,
+                                  size_t number_of_vectors,
+                                  size_t number_of_grid_points) {
+  auto value_distribution = std::uniform_real_distribution(
+      std::numeric_limits<double>::min(), std::numeric_limits<double>::max());
+
+  std::vector<T> res(number_of_vectors);
+  for (size_t i = 0; i < number_of_vectors; ++i) {
+    res[i] = make_with_random_values<T>(gen, value_distribution,
+                                        number_of_grid_points);
+  }
+  return res;
+}
+
+template <typename T>
+void check_constructors() {
+  MAKE_GENERATOR(gen);
+  auto size_distribution = std::uniform_int_distribution(1, 10);
+  auto number_of_grid_points = static_cast<size_t>(size_distribution(gen));
+  const auto number_of_vectors = static_cast<size_t>(size_distribution(gen));
+  if constexpr (std::is_same_v<double, T>) {
+    number_of_grid_points = 1;
+  }
+
+  CAPTURE(number_of_grid_points);
+  CAPTURE(number_of_vectors);
+
+  const auto expected = create_random_data<T>(
+      make_not_null(&gen), number_of_vectors, number_of_grid_points);
+
+  auto dynamic_buffer = std::make_unique<DynamicBuffer<T>>(
+      number_of_vectors, number_of_grid_points);
+
+  for (size_t i = 0; i < number_of_vectors; ++i) {
+    dynamic_buffer->at(i) = expected.at(i);
+  }
+
+  serialize_and_deserialize(*dynamic_buffer);
+  check_results(*dynamic_buffer, expected);
+
+  test_copy_semantics(*dynamic_buffer);
+  auto dynamic_buffer_for_move = *dynamic_buffer;
+  test_move_semantics(std::move(dynamic_buffer_for_move), *dynamic_buffer);
+
+  DynamicBuffer<T> dynamic_buffer_copied(*dynamic_buffer);
+  DynamicBuffer<T> dynamic_buffer_assigned;
+  dynamic_buffer_assigned = *dynamic_buffer;
+
+  if constexpr (tt::is_iterable_v<T>) {
+    auto copied_it = dynamic_buffer_copied.begin();
+    auto assigned_it = dynamic_buffer_assigned.begin();
+    // check that the new buffers don't point to the same data storage location
+    for (auto it = dynamic_buffer->begin(); it != dynamic_buffer->end();
+         ++it, ++copied_it, ++assigned_it) {
+      // first `DataVector` of dynamic_buffer
+      const auto first_data_vector = *it;
+
+      // underlying C array
+      const auto array = first_data_vector.data();
+
+      // address of first element
+      const auto address = &*array;
+
+      // analogously for copied DynamicBuffers
+      CHECK(&*((*copied_it).data()) != address);
+      CHECK(&*((*assigned_it).data()) != address);
+    }
+    CHECK(copied_it == dynamic_buffer_copied.end());
+    CHECK(assigned_it == dynamic_buffer_assigned.end());
+  }
+
+  DynamicBuffer<T> dynamic_buffer_moved = std::move(*dynamic_buffer);
+
+  dynamic_buffer.reset();  // calls destructor
```

Comment:
Why are you doing this?

Reply:
I have removed it since we are already checking that the memory addresses are not equal for each component of the copied/moved buffers.